### PR TITLE
adding gallery sorting

### PR DESCRIPTION
**Pull request by choldgraf**
Now that sphinx-gallery allows you to choose folder ordering within galleries, this is a PR that attempts to give this a shot. I figured it'd be cumbersome to choose a custom ordering for *all* of our folders, since really we just wanna make sure that a subset are at the top. So, this PR defines that subset, and then lists all the subsequent folders alphabetically.

This PR does:

1. Add gallery ordering (with a semi-arbitrary ordering chosen by me...lmk if you think it should be different)
2. Rename the tutorials folders from e.g., `01_introductory` -> `introductory` since we don't need alphanumeric ordering anymore

Let's see how it looks in circle!

### CODE REVIEW

File: doc/conf.py [hunk 1]
```diff
@@ -95,6 +97,18 @@ def _check_deps():
...
 
+explicit_order_folders = ['../examples/pyplots',
+                          '../examples/statistics',
+                          '../examples/color',
+                          '../examples/pylab_examples',
+                          '../examples/showcase',
+                          '../tutorials/introductory',
+                          '../tutorials/intermediate',
+                          '../tutorials/advanced']
+for folder in glob('../examples/*') + glob('../tutorials/*'):
```
Comment (QuLogic) [inline doc/conf.py]: `glob` is not sorted.

**Comment by choldgraf** [comment]
ok I think I got all the cross references this time :-)

File: doc/conf.py [hunk 2]
```diff
@@ -95,6 +97,18 @@ def _check_deps():
   'pandas': ('http://pandas.pydata.org/pandas-docs/stable', None)
   }
 
+explicit_order_folders = ['../examples/pyplots',
+                          '../examples/statistics',
+                          '../examples/color',
+                          '../examples/pylab_examples',
```
Comment (QuLogic) [inline doc/conf.py]: I think this just got removed with #8983.

File: doc/conf.py [hunk 3]
```diff
@@ -95,6 +97,18 @@ def _check_deps():
...
+                          '../examples/showcase',
```
Comment (QuLogic) [inline doc/conf.py]: Showcase should maybe be higher?

File: doc/conf.py [hunk 4]
```diff
@@ -95,6 +97,18 @@ def _check_deps():
...
+                          '../tutorials/introductory',
```
Comment (QuLogic) [inline doc/conf.py]: Where should I be looking to find this? In the gallery page, I don't think I see any tutorials after showcase?

Comment (choldgraf) [inline doc/conf.py]: The way SG handles this is there's just one list that defines the order priority. So folders inside of tutorials and examples won't be shown together or anything, this just defines the ordering. You can think of this as two separate lists, one for examples and one for tutorials.

File: doc/conf.py (hunk 3 above)
Comment (choldgraf) [inline doc/conf.py]: +1 for that...any others you'd like to see reordered?

File: doc/conf.py (hunk 2 above)
Comment (choldgraf) [inline doc/conf.py]: whoah way to go @NelleV ! I didn't know we were so close to moving everything outta that folder. I'll rebase and see if it's gone now.

File: doc/conf.py (hunk 3 above)
Comment (choldgraf) [inline doc/conf.py]: Actually - I just realized that a better solution here might be to use the viewing data from google analytics to choose the ordering. That might be a kind of self-reinforcing kind of thing, but it's a good first start

**Comment by story645** [comment]
Happy dance that there's now ordering, but I do wonder if there's some natural clustering of the categories aside from metrics. The list is:

> Pyplot Examples
> Statistics
> Color
> Pylab Examples
> Showcase
> Animation Examples
> Matplotlib API
> Axes Grid
> Axis Artist
> Event Handling
> Front Page
> Images, contours and fields
> Lines, bars and markers
> Miscellaneous Examples
> mplot3d toolkit
> Pie and polar charts
> Our Favorite Recipes
> Scales in Matplotlib
> Shapes and collections
> Specialty Plots
> Style sheets
> Subplots, axes and figures
> Text, labels and annotations
> Ticks and spines
> Units in Matplotlib
> Embedding matplotlib in graphical user interfaces
> Userdemo
> Widgets

I expect that some of these are probably are already being collapsed into each other, but I think some limit like n=10 or whatever a usability expert would suggest on categories (there are almost 30 in the list) could help make it all seem less overwhelming?

**Comment by choldgraf** [comment]
So I am definitely +1 on refactoring the examples (I actually think they should fall into a few sub-categories, since many of those examples aren't going to be of interest to non-dev users). But - that'll be complicated and will do things like breaking hard-links, so I'd prefer if we kept this PR focused on picking and implementing an order. WDYT?

**Comment by story645** [comment]
I totally agree it's out scope for this PR, it just seems like the sort of thing someone is already working on concurrently and so figured I'd throw it out there. 

As a first pass on ordering, I'm thinking roughly the following clusters:

> Pyplot Examples (while this is most searched for, I debate promoting it to top since this isn't the recommended style)

> Statistics
> Pylab Examples
> Front Page
> Showcase
> Miscellaneous Examples
> Our Favorite Recipes
> Specialty Plots

> Matplotlib API (though should maybe be moved right under/above pyplot examples)
> Animation Examples
> Axes Grid
> Axis Artist
> Color
> Images, contours and fields
> Lines, bars and markers
> Pie and polar charts
> Shapes and collections
> Scales in Matplotlib
> Subplots, axes and figures
> Style sheets
> Text, labels and annotations
> Ticks and spines
> Units in Matplotlib
> mplot3d toolkit

> Event Handling
> Embedding matplotlib in graphical user interfaces
> Widgets
> Userdemo

**Comment by WeatherGod** [comment]
That first group just seems like a weird mishmash. I mean, what are
"Specialty Plots" that aren't "Favorite Recipes". And why wouldn't our
favorites be "Showcased"?
 And Statistics stands out like a sore thumb in that group. This is more an
argument to possibly rename those sections and/or consolidate them further.

**Comment by story645** [comment]
Totally a mishmash. I just grabbed all the categories that just seem to be about displaying cool plots rather than showing off specific parts of the API 'cause they didn't seem to fit in anywhere else. I totally think they should be renamed and consolidated further.

**Comment by choldgraf** [comment]
Just to re-iterate, this PR isn't about renaming or combining example folders...just agreeing on an initial ordering, though I think it's a good idea to open a subsequent PR about combining etc. I think if that's a direction people are friendly to, it should happen before 2.1 is released, because this will break a lot of links.

to @story645's ordering, I think this is a good start! Agree we should shuffle things around a little bit. @WeatherGod do you have specific changes to the order that make more sense to you?

**Comment by choldgraf** [comment]
Here's a quick plot of the top ~20 pages inside of examples. I think we should assume pylab will be split up (which is a great idea IMO).

image

**Comment by dstansby** [comment]
@choldgraf pylab_examples is no more in the `master` branch

**Comment by choldgraf** [comment]
ok latest push no longer has pylab examples!

**Comment by WeatherGod** [comment]
I would put Statistics and Miscellaneous as the last two in the first group.

**Comment by tacaswell** [comment]
:+1: to getting the sorting in at all and defer bikeshedding of names, organization, etc to later work.

**Pull request merged**